src/azure-cli-core/azure/cli/core/commands/__init__.py

```diff
@@ -199,12 +199,25 @@ def execute(self, args):
                                   EVENT_INVOKER_FILTER_RESULT)
         from knack.util import CommandResultItem, todict
         from azure.cli.core.commands.events import EVENT_INVOKER_PRE_CMD_TBL_TRUNCATE
+        from azure.cli.core.commands import cache
 
         # TODO: Can't simply be invoked as an event because args are transformed
         args = _pre_command_table_create(self.cli_ctx, args)
 
         self.cli_ctx.raise_event(EVENT_INVOKER_PRE_CMD_TBL_CREATE, args=args)
-        self.commands_loader.load_command_table(args)
+        success = False
+        if cache.cache_exists():
+            logger.debug("attempting to use cache")
+            success = cache.load_command_table(self.commands_loader, args) is not None
+
+        if not success or not cache.cache_exists():
+            logger.debug("not using cache")
+            self.commands_loader.load_command_table(args)
+            # only cache if there is a a cmd_to_loader_map, handles "TestCommandsLoader"
+            if os.getenv('AZ_USE_CACHE', 'False').lower() == 'true' and hasattr(self.commands_loader, 'cmd_to_loader_map'):  # pylint: disable=line-too-long
+                logger.debug("persisting cache")
+                cache.cache_command_table(self.cli_ctx, self.commands_loader.cmd_to_loader_map)
```

Comment:
It would make more sense for the method to contain the debug statement. I would recommend including the path to the cached file in the debug statement.

Reply:
I would simplify this flow:
```Python
if USE_CACHE:
  try:
     cache.load_command_table(…)
  except CacheNotExistError:
     pass

if not self.commands_loader.command_table:  # assuming that loading the cache populates the command table
   self.commands_loader.load_command_table(args)

if USE_CACHE:
   cache.cache_command_table(…)
```